compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2772,6 +2772,32 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                         ),
                         _ => return None,
                     };
+
+                    // Heuristically determine whether a type name is likely intended to be a generic.
+                    //
+                    // We apply three rules:
+                    // 1. Short names (like `T`, `U`, `E`) are common for generics.
+                    // 2. Certain well-known names (e.g., `Item`, `Output`, `Error`) are commonly used as generics.
+                    // 3. Names in UpperCamelCase are more likely to be concrete types.
+                    //
+                    // This approach may produce false positives or negatives, but works well in most cases.
+
+                    let common_generic_names: &[&str] = &[
```

Comment:
```suggestion
                    let common_type_param_names: &[&str] = &[
```

since it doesn't make sense for const params

Reply:
Also, I'd like to draw attention to the

```rust
    && segment.ident.name != kw::SelfUpper
    && segment.ident.name != kw::Dyn =>
```

at the very start of the function. We're already filtering out certain identifiers (in this case `Self` and `dyn`). I'd like us to unify this check and the `common_generic_names` one is some way or another

Reply:
this is good point, will it be better to replace my logic right after this `let (ident, span) = ...` or integrate them inside the `match`?

Reply:
I think what @fmease might expect is a change like...

```rust

        let [segment] = segments else {
               return None;
         };

          // comment here explaining why we skip these
         let skip_names = [kw::SelfUpper, kw::Dyn, sym::This, sym::Error, /* etc */];

         if !segment.has_generic_args && !skip_names.contains(segment.ident.name) {
                 return None;
         }
         let ident = segment.ident.to_string();
         let span =  segment.ident.span;
```

I personally much prefer let_else over a match in this case. Feel free to write it with a match if you prefer.

Also, not all these names are available as symbols in `sym`. Feel free to add them there as needed,